ten-simple-rules-dockerfiles.Rmd

```diff
@@ -109,22 +113,36 @@ In fact, the `Dockerfile` itself can be a powerful documentation to show from wh
-These tools can add required software packages to an existing image without you having to manually write a `Dockerfile` at all.
-Well-maintained includes the tool's own stability and usability, but also that suitable base images are used, likely from the official Docker library [@docker_inc_official_2019], to ensure that the container has the most recent security fixes for the operating system in question.
-Third, you may want to write you own `Dockerfile` from scratch if these tools do not suffice for your needs.
-_In this case, follow the remaining rules._
+Rule 1 could informally be described as "Don't bother to write a
+Dockerfile!"  Writing a `Dockerfile` from scratch can be difficult
+and even experts sometimes take shortcuts.  Thus, it is a good
+strategy to first look to tools that can help to generate a
+`Dockerfile` for you.  Such tools have likely thought about and
+implemented good practices, and they may have added newer practices
+when reapplied at a later point in time.  Therefore the most important
+rule is to apply a multi-step process for your specific use case.
+First, you want to determine if there is an already existing container
+that you can use, and in this case, use it and add to your workflow
+documentation instructions for doing so.  As an example, you might be
+doing some kind of interactive development.  For interactive
+development environments such as notebooks and development servers or
+databases, you can readily find containers [SJE: where do you look to find containers?] that come installed with
+all the software that you need.  Second, in the case that there isn't
+an existing container for your needs, you might next look to
+well-maintained tools to help with `Dockerfile` generation.  These
+tools can add required software packages to an existing image without
+you having to manually write a `Dockerfile` at all.  Well-maintained
+includes the tool's own stability and usability, but also that
+suitable base images are used, likely from the official Docker library
+[@docker_inc_official_2019], to ensure that the container has the most
+recent security fixes for the operating system in question.  Third,
+you may want to write you own `Dockerfile` from scratch if these tools
+do not suffice for your needs.  _In this case, follow the remaining
+rules._
 
 ## Tools for container generation
 
-Repo2docker [@jupyter_binder_2018] is a tool that is maintained by [Project Jupyter](https://jupyter.org/) that can help to transform a repository with common simple configuration files for defining software dependencies and versioninto a container.
+Repo2docker [@jupyter_binder_2018] is a tool that is maintained by [Project Jupyter](https://jupyter.org/) that can help to transform a [SJE github/gitlab? any other repos?] repository with common simple configuration files for defining software dependencies and versioninto a container.
```

Comment:
From most places that people would use for version control, yes to all of those, see https://repo2docker.readthedocs.io/en/latest/

Reply:
> One quick global comment - we need to establish early on the scope (small data science scripts) of this paper. We cannot open with a flowery and misleading statement about reproducible workflows because we don’t get anywhere near discussing scale, workflow execution, and generally more complex (real world) pipelines that use multiple containers, might interact with TB of data, and be run at scale with some cloud provider or HPC job manager or similar. We don’t want someone reading this that falls into that boot camp to roll their eyes because they intuit that the authors don’t have a clue about this scale of work.

agree re: this.  Perhaps we could have a section at the start regarding a typical reader that might benefit from these tips?

Reply:
Awesome! @nuest if you want to ping me when the rounds of edits / changes are done, I can do a final read over _and_ I'll write this scoped section, which I think is super important. I'll open an issue and you can assign me to it so that we don't forget :)

Reply:
https://github.com/nuest/ten-simple-rules-dockerfiles/issues/55